addons/html_builder/static/src/core/builder_selection_restriction_plugin.js

```diff
@@ -0,0 +1,242 @@
+    }
+
+    restrictSelectionInElement(selection, block) {
+        const { anchorNode, anchorOffset, focusNode, direction } = selection;
+        const isFocusInBlock = block.contains(focusNode);
+
+        if (!isFocusInBlock) {
+            let focusNode, focusOffset;
+            if (direction === DIRECTIONS.RIGHT) {
+                [focusNode, focusOffset] = getDeepestPosition(block, nodeSize(block));
+            } else {
+                [focusNode, focusOffset] = getDeepestPosition(block, 0);
+            }
+            this.dependencies.selection.setSelection({
+                anchorNode,
+                anchorOffset,
+                focusNode,
+                focusOffset,
+            });
+        }
+        this.correctSelectionOnUncrossable();
+    }
+
+    restrictSelectionInClosestDiv(ev) {
+        const { editableSelection, currentSelectionIsInEditable } =
+            this.dependencies.selection.getSelectionData();
+        if (!currentSelectionIsInEditable) {
+            return;
+        }
+        if (editableSelection.isCollapsed) {
+            return;
+        }
+
+        const closestSpecialBlock = closestElement(
+            editableSelection.anchorNode,
+            this.special_block_with_text_in_non_div.join(",")
+        );
+        const closestDiv = closestElement(editableSelection.anchorNode, "div");
+
+        if (closestSpecialBlock && !closestSpecialBlock.contains(closestDiv)) {
+            const closestParagraph = closestElement(editableSelection.anchorNode, "p");
+            this.restrictSelectionInElement(editableSelection, closestParagraph);
+        } else {
+            this.restrictSelectionInElement(editableSelection, closestDiv);
+        }
+    }
+
+    isNodeSelectionUncrossable(node, selectedNodes) {
+        return (
+            node.matches(this.uncrossable_element_selector) ||
+            selectedNodes.includes(closestElement(node, this.uncrossable_element_selector))
+        );
+    }
+
+    correctSelectionOnUncrossable() {
+        const selection = this.dependencies.selection.getEditableSelection();
+        const { anchorNode, anchorOffset, focusNode, direction } = selection;
+        const selectedNodes = this.dependencies.selection
+            .getTargetedNodes(selection)
+            .filter((node) => node.nodeType === Node.ELEMENT_NODE);
+        const selectedNodesLooping =
+            direction === DIRECTIONS.RIGHT ? selectedNodes : selectedNodes.reverse();
+
+        // do not do selection correction if 1. the only selected node is an image cause we force
+        // the selection to be around the image when clicking on it, correction would break the
+        // option container for images 2. icon elements (i.fa, span.fa...)
+        if (
+            selectedNodesLooping.length === 1 &&
+            (selectedNodesLooping[0].tagName === "IMG" ||
+                selectedNodesLooping[0].classList?.contains("fa"))
+        ) {
+            return;
+        }
```

Comment:
- Improve the comment like this: 
  ```js
  // Do not correct the selection if the only selected node is an image or
  // an icon, otherwise it would break their option containers (because we
  // force the selection to be around them when we click on them).
  ```
- Can we generalize this to all media ? In that case, I would do this
  ```js
   if (selectedNodes.length === 1 && isMediaElement(selectedNodes[0]))
  ```
  Otherwise, at least do this to use the `isIconElement` utils :slightly_smiling_face: 
  ```js
    if (
      selectedNodes.length === 1 &&
      (selectedNodes[0].tagName === "IMG" || isIconElement(selectedNodes[0]))
    )
  ```